runtime/runtime-core/jvm/src/aws/smithy/kotlin/runtime/io/GzipSdkSource.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.InternalApi
+import java.util.zip.GZIPOutputStream
+
+@InternalApi
+public class GzipSdkSource(
+    private val source: SdkSource,
+    private val bytesAvailable: Long?,
+) : SdkSource {
+    private val gzipBuffer = SdkBuffer()
+    private val gzipOutputStream = GZIPOutputStream(gzipBuffer.outputStream(), true)
+    private var bytesRead: Long = 0
+
+    /**
+     * Keeps track of whether a read operation has been made on this sdk source
+     */
+    private var read: Boolean = false
+
+    override fun read(sink: SdkBuffer, limit: Long): Long {
```

Comment:
comment: In general these `read` functions seem complicated, I'll have to pull it down and take a longer look but I feel it could be simplified

Reply:
Agreed, these need simplified. See my previous comment https://github.com/awslabs/smithy-kotlin/pull/1000#discussion_r1407933393